python/asrv/proto.py

```diff
@@ -0,0 +1,368 @@
+        for qid, sub in self.subs.items():
+            sub.kill()
+        self.srv.remove_conn(self)
+
+    def send_msg(self, msg):
+        self.transport.write(self.packer.pack(msg))
+
+    def getnint(self, msg, field):
+        val = msg.get(field, None)
+        if not isinstance(val, int) and val is not None:
+            raise ProtocolError('{} is not int or None'.format(field))
+        return val
+
+    def getid(self, msg, field):
+        val = msg.get(field, None)
+        if val is None:
+            val = bytes(24)
+        if not isinstance(val, bytes) or len(val) != 24:
+            raise ProtocolError('{} is not a valid id'.format(field))
+        if val == bytes(24):
+            val = None
+        return val
+
+    def getfbool(self, msg, field):
+        val = msg.get(field, False)
+        if not isinstance(val, bool):
+            raise ProtocolError('{} is not bool'.format(field))
+        return val
+
+    def msg_create(self, msg):
+        parent = self.getid(msg, 'parent')
+        pos_start = self.getnint(msg, 'pos_start')
+        pos_end = self.getnint(msg, 'pos_end')
+        tags = msg.get('tags', [])
+        attr = msg.get('attr', {})
+        data = msg.get('data', {})
+        bindata = msg.get('bindata', {})
+        if not isinstance(tags, (list, tuple)):
+            raise ProtocolError('tags is not list')
+        for tag in tags:
+            if not isinstance(tag, str):
+                raise ProtocolError('tag is not string')
+        tags = set(tags)
+        if not isinstance(attr, dict):
+            raise ProtocolError('attr is not dict')
+        for key in attr:
+            if not isinstance(key, str):
+                raise ProtocolError('attr is not string')
+        if not isinstance(data, dict):
+            raise ProtocolError('data is not dict')
+        for key in data:
+            if not isinstance(key, str):
+                raise ProtocolError('data key is not string')
+        if not isinstance(bindata, dict):
+            raise ProtocolError('bindata is not dict')
+        for key, val in bindata.items():
+            if not isinstance(key, str):
+                raise ProtocolError('bindata key is not string')
+            if not isinstance(val, bytes):
+                raise ProtocolError('bindata val is not bytes')
+        obj_id = msg.get('id')
+        if not isinstance(obj_id, bytes) or len(obj_id) != 24:
+            raise ProtocolError('obj_id is not bytes with len 24')
+        aid = self.getnint(msg, 'aid')
+        self.srv.create(
+            obj_id,
+            parent,
+            tags=tags,
+            attr=attr,
+            data=data,
+            bindata=bindata,
+            pos=(pos_start,pos_end),
+        )
+        if aid is not None:
+            self.send_msg({
+                'type': 'ack',
+                'aid': aid,
+            })
+
+    def msg_modify(self, msg):
+        # XXX
+        raise NotImplementedError
```

Comment:
Hmm, throw some error that will be caught and send NotImplementedError message to client, instead?